rust/lance-linalg/src/distance/cosine.rs

```diff
@@ -89,7 +91,76 @@ where
 
 impl Cosine for BFloat16Type {}
 
-impl Cosine for Float16Type {}
+#[cfg(any(
+    all(target_os = "macos", target_feature = "neon"),
+    all(target_os = "linux", feature = "avx512fp16")
+))]
+mod kernel {
+    use super::*;
+
+    extern "C" {
+        pub fn dot_f16(ptr1: *const f16, ptr2: *const f16, len: u32) -> f32;
```

Comment:
shall we move `dot` and `norm_l2` kernel to their own files?

Reply:
Removed.